### ISSUE TITLE
Delay until song rating is transferred to Apple Music

### ISSUE THREAD

**Issue by thekryz** [MainasuK/Song-Rating#18]
Hi there! Thank you for your app, I use it a lot!
I have a huge unrated library of music, so I rate a lot. Often, I want to rate songs quickly and jump to the next one. But there is a delay - and checking your code, it's there deliberately to "debounce" (?) for 2s in the setRating function. Is this really (still) necessary? I would appreciate an immediate (or only slightly) delayed rating transferral. Would removing the debouncing be possible or what issues does it lead to?

**Comment by thekryz** [comment]
How about something along these lines @ iTunesRadioStation.swift:

```
class LeadingEdgeDebouncer {
    private let delay: TimeInterval
    private var workItem: DispatchWorkItem?
    private var lastExecutionTime: Date?

    init(delay: TimeInterval) {
        self.delay = delay
    }

    func debounce(_ block: @escaping () -> Void) {
        // Cancel any pending execution
        workItem?.cancel()

        let now = Date()
        if lastExecutionTime == nil || now.timeIntervalSince(lastExecutionTime!) > delay {
            // Execute immediately if it's the first call or if enough time has passed since the last execution
            lastExecutionTime = now
            block()
        } else {
            // Debounce subsequent calls
            let newWorkItem = DispatchWorkItem {
                self.lastExecutionTime = Date()
                block()
            }
            workItem = newWorkItem
            // Schedule the work item to be executed after the specified delay
            DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: newWorkItem)
        }
    }
}

extension iTunesRadioStation {
    // Create a debouncer with a 0.3 second delay
    private let ratingDebouncer = LeadingEdgeDebouncer(delay: 0.3)

    func setRating(_ rating: Int) {
        // Invalidate any existing timer
        debounceSetRatingTimer?.invalidate()

        // Ensure we have a current track to rate
        guard latestPlayInfo != nil || !(iTunes?.currentTrack?.name ?? "").isEmpty else {
            os_log("%{public}s[%{public}ld], %{public}s: try to set rating but no current track info", ((#file as NSString).lastPathComponent), #line, #function)
            return
        }

        let name = latestPlayInfo?.name ?? iTunes?.currentTrack?.name ?? "nil"
        os_log("%{public}s[%{public}ld], %{public}s: debouncing set rating for %{public}s %{public}ld…", ((#file as NSString).lastPathComponent), #line, #function, name, rating)

        // Use the debouncer to handle the rating change
        ratingDebouncer.debounce { [weak self] in
            guard let self = self else { return }
            let track = self.iTunes?.currentTrack
            // Set the rating on the track
            track?.setRating?(rating)
            os_log("%{public}s[%{public}ld], %{public}s: … set %{public}s rating %{public}ld", ((#file as NSString).lastPathComponent), #line, #function, track?.name ?? "nil", rating)
        }
    }
}
```

this should implement a Leading Edge Debouncer which will:
1. Apply the first rating change immediately
2. Debounce subsequent rapid changes within a short time window (e.g., 0.3 seconds)

This way the app would instantly transfer the initial rating while protecting against potential issues from rapid successive changes.

**Comment by MainasuK** [comment]
This delay may cause the rating action to execute on the next track when playback at the end of the song. The track object should be a deep copy (not the current shadow copy) to keep the point to the track on which the rating action on. 

I will try to fix this issue with the patch for the new macOS update. Thank you for your bug report.

**Comment by MainasuK** [comment]
Should be fixed in v1.5.0. The delay rating trigger after the song playback ends will be set to the correct one now.

```
iTunesRadioStation.swift[181], setRating(_:): set timer for 2.0s and set rating for 夙春殇  40…
iTunesRadioStation.swift[139], playInfoChanged(_:): Playing: [2] Nightlight | 小明是根草 feat. 洛天依 | 平行四界VI (2018) | -1
iTunesPlayerHistory.swift[37], insert(_:): playback history append track: Nightlight
RatingControl.swift[80], update(rating:): draw rating control 0
iTunesRadioStation.swift[191], setRating(_:): set rating for 夙春殇 : 40
```
